addons/controllers/csi/vspherecsiconfig_utils.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	csiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/csi/v1alpha1"
+)
+
+// mapVSphereCSIConfigToDataValues maps VSphereCSIConfig CR to data values
+func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValues(ctx context.Context,
+	vcsiConfig *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (*DataValues, error) {
+	switch vcsiConfig.Spec.VSphereCSI.Mode {
+	case VSphereCSINonParavirtualMode:
+		return r.mapVSphereCSIConfigToDataValuesNonParavirtual(ctx, vcsiConfig, cluster)
+	case VSphereCSIParavirtualMode:
+		return r.mapVSphereCSIConfigToDataValuesParavirtual(ctx, vcsiConfig, cluster)
+	default:
+		break
+	}
+	return nil, errors.Errorf("Invalid CSI mode '%s', must either be '%s' or '%s'",
+		vcsiConfig.Spec.VSphereCSI.Mode, VSphereCSIParavirtualMode, VSphereCSINonParavirtualMode)
+}
+
+func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesParavirtual(ctx context.Context,
+	vcsiConfig *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (*DataValues, error) {
+	dvs := &DataValues{}
+	config := vcsiConfig.Spec.VSphereCSI.ParavirtualConfig
+
+	if config == nil {
+		return nil, errors.Errorf("ParavirtualConfig missing in Spec.VSphereCSI")
+	}
+
+	// first populate derived values
+	dvs.VSpherePVCSI = &DataValuesVSpherePVCSI{}
+
+	dvs.VSpherePVCSI.Namespace = tryParseClusterVariableString(ctx, cluster, NamespaceVarName)
```

Comment:
question: could we use `res, err := util.ParseClusterVariableString(cluster, variableName)` directly ? Trying to educate myself on what benefits there are to have a wrapper `tryParseClusterVariableString()`

Reply:
this is better because we don't have to test for 'err' multiple times then set an empty string over and over again.